### ISSUE TITLE
Implement a Golang client 

### ISSUE THREAD

**Issue by AlvinKuruvilla** [sirixdb/sirix#361]
### Feature idea
I think it would be a good idea to implement a Go client, If there is interest I would probably go off both the Python and /or Rust client. I'm new to Go so it will probably take me a while, but I think this will be a good experience and also give me a way to contribute to this great repo

**Comment by JohannesLichtenberger** [comment]
@AlvinKuruvilla, that's great :-) let me and @mosheduminer know if you need help. You can have a look into the [Python-client](https://github.com/sirixdb/sirix-python-client).

**Comment by AlvinKuruvilla** [comment]
Thanks @JohannesLichtenberger 
Here is the link to the GitHub repo, it's still kind of barebones but the basic structure should mostly be in place
https://github.com/AlvinKuruvilla/Sirix-Go

**Comment by mosheduminer** [comment]
Excellent @AlvinKuruvilla! I'm following development 🙂.

**Comment by nbrugger-tgm** [comment]
As far as I understand you mean a Go lib accessing the ReST api?

In this case the client might even be easy to auto-generate using techniques like open api generator depending on the API specification standard used (RAML, OAS, blueprint). This way it would be low to no effort to keep many different client languages up to date and also to support many languages like nodejs , python, Java, go, c++ and c#

**Comment by mosheduminer** [comment]
> depending on the API specification standard used

Currently we have no machine readable specification document, and we would need to write it by hand as well (due to the nature of vertx). So in theory we could use any.

However - 

> the client might even be easy to auto-generate

How do these generated clients handle auth? The biggest (perhaps only) challenge when implementing the client is auth. I don't want the client to need to retrieve credentials before every request. In addition, we must consider connection pooling, as well as how well the generator keeps the paradigm of the language (for example, we probably don't want to parse JSON or XML into classes in python).

**Comment by JohannesLichtenberger** [comment]
@nbrugger-tgm, I guess it would be great to have a Swagger-file and auto-generate clients. However, I also had some trouble with authentication / authentification when I once did some tests regarding another project / API.

**Comment by nbrugger-tgm** [comment]
@mosheduminer
> Currently we have no machine readable specification document, and we would need to write it by hand as well (due to the nature of vertx). So in theory we could use any.

Yes and no, you have to **do** it by hand but not **writing** it per had. As for OAS/Swagger there are quite a few WYSIWYG editors such as Stoplight Studio (free). I would go with OAS but that might be biased

> How do these generated clients handle auth?...

Well in the case of OAS you have different options, (OAuth, Bearer) and API tokens where you just define a (querry, path or header) parameter that has to be present on certain requests. Also if this is a big problem you might do changes to the auto-generated clients.

Regarding the other concerns, this sole relies on the generator used. `openapi-generator` for example uses templates that are built into it, but as the project is open source you can easily fork the generator and adapt the templates for your need or even create custom ones.


@JohannesLichtenberger i think having a swagger/oas file is also very advisable if you decide against using the client-auto generation. The reasons are pretty clear i think, starting at a documentation/specification pov.


How is the authentication on the rest api working that you think i might not works as you want it / inconvenient?

**Comment by JohannesLichtenberger** [comment]
@nbrugger-tgm I think it probably isn't an issue because of our JWT token based auth. At work I had some trouble getting the session-ID stuff right and the generated Java client was a bit "ugly" in comparison to "hand-written" code. But I think it would be a great contribution to create an OpenAPI based spec and auto-generate clients as for instance for our main language Java (as well as Kotlin) :+1:

You can find the API description over here: https://sirix.io/rest-api.html#api-description

**Comment by AlvinKuruvilla** [comment]
Yeah, that's true @JohannesLichtenberger. In that vein, any assistance on the Golang client would be greatly appreciated. It sort of fell to the side for me due to other projects and general life. There is nothing in particular but more of just general contributions and more Go paradigm style coding. The OpenAPI idea sounds really cool I am going to see if there is anything I can contribute but this is new territory for me
